### DISCUSSION TITLE
Hide stack info when LOG_LEVEL warn or higher

### DISCUSSION THREAD

**Discussion by ybelenko** [directus/directus#11724] [Archive]
I would hide error stack field from API response when `LOG_LEVEL` isn't debug.

### To Reproduce

Create new blank extension with the `src/index.js`:
```js
export default ({ filter, action }, { exceptions }) => {
	const { InvalidPayloadException } = exceptions;
	filter('items.create', () => {
		throw new InvalidPayloadException('Foobar');
		console.log('Creating Item!');
	});

	action('items.create', () => {
		console.log('Item created!');
	});
};
```

### Errors Shown

HTTP POST: /collection to raise that validation error.
```json
{
  "errors": [
    {
      "message": "Foobar",
      "extensions": {
        "code": "INVALID_PAYLOAD",
        "stack": "Error: Foobar\n    at /Users/ybelenko/Sites/directus/api/extensions/hooks/issue11275/index.js:1:132\n    at Emitter.emitFilter (/Users/ybelenko/Sites/directus/api/src/emitter.ts:36:25)\n    at /Users/ybelenko/Sites/directus/api/src/services/items.ts:100:22\n    at /Users/ybelenko/Sites/directus/node_modules/knex/lib/execution/transaction.js:221:22\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)"
      }
    }
  ]
}
```
Again field from `errors[0].extensions.stack` with line breaks:
> Error: Foobar
> at /Users/ybelenko/Sites/directus/api/extensions/hooks/issue11275/index.js:1:132
> at Emitter.emitFilter (/Users/ybelenko/Sites/directus/api/src/emitter.ts:36:25)
> at /Users/ybelenko/Sites/directus/api/src/services/items.ts:\100:22
> at /Users/ybelenko/Sites/directus/node_modules/knex/lib/execution/transaction.js:221:22
> at processTicksAndRejections (node:internal/process/task_queues:96:5)"

If I set `LOG_LEVEL` env value to `silent` I still see numbers of code lines and file names. 9 of 10 security audits would report that behaviour as low level vulnerability.

**Comment by rijkvanzanten** [comment]
The stack is only included for admin users. We'll treat this discussion as a feature request to add a new flag to enable/disable that 👍🏻

**Reply by ybelenko** [reply]
> The stack is only included for admin users. We'll treat this discussion as a feature request to add a new flag to enable/disable that 👍🏻

Can I apply some `accountability` in my custom extension to hide this stack property? My collection is public and open to unauthorized POST requests, so I don't send any authorization tokens or something.

**Reply by joselcvarela** [reply]
Can you check if you are making the POST request as an admin?
Just clear all the cookies from you Directus instance and try again.

**Reply by ybelenko** [reply]
I make request from `Insomnia` desktop app. Just deleted all cookies. No authorization headers. `feedback` collection has public create permission.

Insomnia request log:
```log
* Preparing request to http://localhost:8080/items/feedback
* Current time is 2022-02-21T11:01:36.656Z
* Using libcurl/7.69.1 OpenSSL/1.1.1g zlib/1.2.11 brotli/1.0.7 libidn2/2.1.1 libssh2/1.9.0 nghttp2/1.41.0
* Using default HTTP version
* Enable timeout of 1000000ms
* Enable automatic URL encoding
* Enable SSL validation
* Enable cookie sending with jar of 0 cookies
* Hostname in DNS cache was stale, zapped
*   Trying ::1:8080...
* Connection failed
* connect to ::1 port 8080 failed: Connection refused
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#1)

> POST /items/feedback HTTP/1.1
> Host: localhost:8080
> User-Agent: insomnia/2020.4.0
> Content-Type: application/json
> Accept: application/json
> Content-Length: 113

| {
| 	"first_name": "john",
| 	"last_name": "doe",
| 	"email": ".@mail.ru",
| 	"phone": "868676767676",
| 	"details": null
| }

* upload completely sent off: 113 out of 113 bytes
* Mark bundle as not supporting multiuse

< HTTP/1.1 400 Bad Request
< Access-Control-Allow-Origin: *
< content-security-policy: script-src 'self' 'unsafe-eval';worker-src 'self' blob:;child-src 'self' blob:;img-src 'self' data: blob:;connect-src 'self' https://*;default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;frame-ancestors 'self';object-src 'none';script-src-attr 'none';style-src 'self' https: 'unsafe-inline'
< x-powered-by: Directus
< vary: Origin
< access-control-allow-credentials: true
< access-control-expose-headers: Content-Range
< content-type: application/json; charset=utf-8
< content-length: 556
< etag: W/"22c-2PAuedeAXS4gFqAGu2nWSvkir2w"
< date: Mon, 21 Feb 2022 11:01:36 GMT
< connection: close


* Received 556 B chunk
* Closing connection 1
```

And the response body(formatted by Insomnia, not raw):
```json
{
  "errors": [
    {
      "message": "Foobar",
      "extensions": {
        "code": "INVALID_PAYLOAD",
        "stack": "Error: Foobar\n    at /Users/ybelenko/Sites/directus/api/extensions/hooks/issue11275/index.js:1:132\n    at Emitter.emitFilter (/Users/ybelenko/Sites/directus/api/src/emitter.ts:36:25)\n    at /Users/ybelenko/Sites/directus/api/src/services/items.ts:100:22\n    at /Users/ybelenko/Sites/directus/node_modules/knex/lib/execution/transaction.js:221:22\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)"
      }
    }
  ]
}
```

**Reply by joselcvarela** [reply]
If you think this is a bug, you may want to open an issue 👍

**Reply by ybelenko** [reply]
> If you think this is a bug, you may want to open an issue 👍

Well, I've opened it as an issue at the first place, then @rijkvanzanten converted it to `Ideas` 😆

**Reply by rijkvanzanten** [reply]
Are you running Directus with NODE_ENV=development? That will also cause the stack to show up 🙂

**Reply by ybelenko** [reply]
> Are you running Directus with NODE_ENV=development? That will also cause the stack to show up 🙂

Yes, it seems so. I start Directus with the command:
```console
npm run dev -- --scope directus --scope @directus/app
```

I would check it with different `NODE_ENV`, but I cannot find how to start Directus from sources in production mode. Your doc [Manual](https://docs.directus.io/getting-started/installation/manual/) doesn't describe how to start the app at all.

**Reply by joselcvarela** [reply]
It seems there's some inconsistency on docs. This has already been reported.
Try to see this page instead: https://docs.directus.io/getting-started/installation/cli/

**Reply by rijkvanzanten** [reply]
> but I cannot find how to start Directus from sources in production mode. 

As you're not really supposed to run Directus from the source code in production 👍🏻

**Reply by ybelenko** [reply]
> As you're not really supposed to run Directus from the source code in production 👍🏻

Even if I'm going to use fork of Directus? Still need to publish it with Docker or something?

**Reply by rijkvanzanten** [reply]
That's totally up to you. Once you fork the platform, it's up to you to decide what to do next

**Reply by ybelenko** [reply]
> That's totally up to you. Once you fork the platform, it's up to you to decide what to do next 

Hmmm, don't get it. I fork the platform not because I want to, but because current version doesn't fit my needs(build of extensions is broken, issue submitted). I would love to use vanilla Directus later, I understand that it's easier to upgrade etc. In the meantime could you provide at least basic doc of how to build the app in production mode.

P.S. I think many users will have to use the forks because right now `npm audit` gives very bad security report and you know it.

**Reply by rijkvanzanten** [reply]
Once somebody forks the platform, any and all changes and deployment configurations are done by third parties that we have absolutely no ties to. There's no way we can support all those different permutations of the software.

> P.S. I think many users will have to use the forks because right now npm audit gives very bad security report and you know it.

We apply security patches to packages that are actually used in the production build of Directus as soon as they come out and are reported, similarly, for any other any other potential security issues, we have a [dedicated email and flow in place](https://github.com/directus/directus/blob/main/security.md). 

As for npm audit, it adds a ton of noise when it comes to potential vulnerabilities in (nested) devDependencies. Most of them are false positives. [Dan Abramov](https://mobile.twitter.com/dan_abramov) wrote a great article on the matter: https://overreacted.io/npm-audit-broken-by-design/.

**Reply by ybelenko** [reply]
I don't ask you to support my fork, I just ask how to start server from sources with `NODE_ENV=production` instead of `dev`. That's all. Don't get me wrong, there is a deadline, I can't wait for months until you fix/implement everything I need, it won't work with my clients. Just give me any advice how to start with prod env instead of dev in a fork.

Once again, until #11577 is solved then I cannot build my own extensions to tune your official vanilla version. It's major reason why I have to use a fork.

**Comment by ybelenko** [comment]
Now I can confirm that with `NODE_ENV=production` response body doesn't contain `code` and `stack`:
```json
{
  "errors": [
    {
      "message": "Foobar",
      "extensions": {
        "code": "INVALID_PAYLOAD"
      }
    }
  ]
}
```

This discussion can be closed, kinda. I'm new to GitHub discussion feature 😅 .
